### DISCUSSION TITLE
🙋🏽‍♀️ 🆘 Symfony Scheduler replacing CRON in production for Amazon Linux 2023 🦮

### DISCUSSION THREAD

**Discussion by trulyprivate** [symfony/symfony#54494] [Q&A]
**_tldr_**: How do you setup Symfony Scheduler for production without using CRON? & How on an architectural level does scheduler actually work? Is there a long lived process in PHP?

I've been looking in the documentation and online for a way I can use Symfony Scheduler as replacement for CRON in AL2023 since it doesn't contain a scheduler by default. 

[Amazon Linux 2023 does not contain /etc/cron.d folder](https://repost.aws/questions/QU0wkAYHXZQxyaCFs0VATkuA/amazon-linux-2023-does-not-contain-etc-cron-d-folder)

I'm currently exploring how to leverage the Symfony Scheduler for a production environment, specifically targeting Amazon Linux 2023, which, to my understanding, doesn’t include a default scheduler akin to `/etc/cron.d`. My objective is to find an alternative to CRON for scheduling tasks within this environment. Additionally, I'm keen on understanding the architectural workings of the Symfony Scheduler, particularly how it manages to remain operational over time.

Here’s the context and my specific questions:

1. **Symfony Scheduler as a CRON Replacement**: 
   - Has anyone successfully set up the Symfony Scheduler in a production environment, especially on Amazon Linux 2023, which lacks a conventional CRON setup?
   - Are there specific configurations or alternative methods recommended for integrating Symfony Scheduler in such environments?

2. **Architectural Insights on Symfony Scheduler**:
   - How does the Symfony Scheduler ensure persistent operation? Is there a mechanism, like a long-lived PHP process, that keeps the scheduler active, or does it initiate self-generating threads?
   - Any insights into how tasks are managed and executed internally by the scheduler would be greatly appreciated.

For additional context, I’ve transitioned a command to utilize Symfony Scheduler, aiming for it to run as a scheduled task. Here's how I've configured it:

```php
<?php

namespace App\Command\Site;

use App\Service\RobotService;
use Monolog\Attribute\WithMonologChannel;
use Psr\Log\LoggerInterface;
use Symfony\Component\Console\Attribute\AsCommand;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Command\LockableTrait;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Scheduler\Attribute\AsCronTask;

#[WithMonologChannel('site_battlebots')]
#[AsCommand(
    name: 'site:battlebots-refresh-robots',
    description: 'Refresh robot caches.',
)]
#[AsCronTask(
    expression: '* * * * *',
    method: 'execute',
    transports: 'async'
)]
class BattleBotsRefreshRobots extends Command
{
    use LockableTrait;

    public function __construct(
        private LoggerInterface $site_robotLogger,
        private RobotService $robotService
    ) {
        parent::__construct();
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        if (!$this->lock()) {
            $this->site_robotLogger->warning('Command already running');
            return Command::SUCCESS;
        }

        $this->site_robotLogger->info('Started');
        $stats = $robotService->generatePages();

        $this->site_robotLogger->info('Completed', ['stats' => $stats]);

        return Command::SUCCESS;
    }
}
```

Given the above, I would deeply appreciate any guidance, best practices, or insights into making the Symfony Scheduler work effectively in Amazon Linux 2023, and understanding its under-the-hood mechanics.

Thank you in advance for your time and help!

**Comment by trulyprivate** [comment]
bump ^^

**Comment by SirTomme** [comment]
There is an infinite loop running started via the `messenger:consume` console command. 

See [this post by Filip Horvat](https://medium.com/@fico7489/symfony-scheduler-how-it-really-works-ef5d95409c09) for an explanation.

**Comment by lmarlin76** [comment]
Hello, I confirm you have to use a worker to consume messages.

See [this post](https://medium.com/@laurentmn/forget-definitively-cron-replace-it-with-symfony-scheduler-5bf8782f53a0) you have a detailed guide how to use it.
